core/frontend/src/extension/ExtensionAdmin.ts

```diff
@@ -0,0 +1,130 @@
+import { ExtensionLoader } from "./ExtensionLoader";
+
+/** The Extensions loading system has 3 goals:
+ *   1. Only fetch what is needed when it is required
+ *      1. Load a manifest file
+ *      2. Load the the main module when necessary
+ *   2. Download the needed files from
+ *
+ * 2 ways to load an Extension into the system:
+ *
+ *  1. At build-time provide the function to load both the Extension Manifest and import the main module of the extension.
+ *     The main module must contain the activate() function.
+ *  2. A minimum set of ExtensionLoaderProps that provide enough information to get the manifest
+ *
+ * An Extension can be in 3 different states:
+ *   - Known
+ *      - The Extension Admin has the minimal information needed to fetch the manifest for the Extension.
+ *        knows the Extension exists but has not yet loaded the manifest
+ *   - Installed
+ *      - The Extension has the full manifest loaded and can be executed on the activation events
+ *   - Disabled
+ *      - The Extension has the full manifest but is not currently enabled and will not be executed based on the
+ *        activation events.
+ */
+
+enum ActivationEvents {
+  onStartupApp = "onStartupApp",
+}
+
+/** The Extension Admin controls the list of currently known, loaded and executing an Extension.
+ * Handles the loading of Extensions and maintains a list of the currently loaded Extensions.
+ *
+ * @beta
+ */
+export class ExtensionAdmin {
+  /** The list of places to download an Extension.  */
+  private _extensionLoaders: ExtensionLoader[] = [];
+
+  /** Defines the set of extensions that are currently known and can be invoked during activation events.  */
+  private _installedExtensions: Map<string, LocalExtensionProps> = new Map<string, LocalExtensionProps>();
+
+  /** Fired when an Extension has been added or removed.
+   * @internal
+   */
+  public onStartup = async () => {
+    await this.activateExtensionEvents(ActivationEvents.onStartupApp);
+  };
+
+  public constructor() {
+    IModelApp.onAfterStartup.addListener(this.onStartup);
+  }
+
+  /** Add an ExtensionLoader to the front of the list of extension loaders. Extension loaders are invoked front to back.
+   * @param extensionLoader Extension loader to add
+   */
+  public addExtensionLoaderFront(extensionLoader: ExtensionLoader) {
+    this._extensionLoaders.unshift(extensionLoader);
+  }
+
+  /** Add an ExtensionLoader to the list of extension loaders in use.
+   * @param extensionLoader Extension loader to add
+   */
+  public addExtensionLoader(extensionLoader: ExtensionLoader) {
+    this._extensionLoaders.push(extensionLoader);
+  }
+
+  /** Add an Extension intend to be bundled during compilation.
+   * @param manifestLoader A function that loads the manifest file.
+   * @param mainFunc The main function to be executed upon
+   */
+  public async addBuildExtension(manifestPromise: Promise<any>, mainFunc?: ResolveFunc): Promise<void> {
```

Comment:
whats the reason for accepting an ambiguous function here as opposed to a typed manifest object that the consumer has fetched on their own (or the very least a type promise result)?

Reply:
updated to accept Promise<ExtensionManifest> and removed the extra parser function